worktree_status.go

```diff
@@ -159,3 +163,103 @@ func diffTreeIsEquals(a, b noder.Hasher) bool {
 
 	return bytes.Equal(hashA, hashB)
 }
+
+// Add adds the file contents of a file in the worktree to the index.
+func (w *Worktree) Add(path string) (plumbing.Hash, error) {
+	s, err := w.Status()
+	if err != nil {
+		return plumbing.ZeroHash, err
+	}
+
+	if s.File(path).Worktree == Unmodified {
+		return plumbing.ZeroHash, nil
+	}
+
+	h, err := w.pushFileToStorer(path)
+	if err != nil {
+		return h, err
+	}
+
+	if err := w.addOrUpdateFileToIndex(path, h); err != nil {
+		return h, err
+	}
+
+	return h, err
+}
+
+func (w *Worktree) pushFileToStorer(filename string) (plumbing.Hash, error) {
+	fi, err := w.fs.Stat(filename)
+	if err != nil {
+		return plumbing.ZeroHash, err
+	}
+
+	f, err := w.fs.Open(filename)
+	if err != nil {
+		return plumbing.ZeroHash, err
+	}
+
+	defer ioutil.CheckClose(f, &err)
+
+	h := plumbing.NewHasher(plumbing.BlobObject, fi.Size())
+	if _, err := io.Copy(h, f); err != nil {
+		return plumbing.ZeroHash, err
+	}
+
+	return h.Sum(), nil
+}
+
+func (w *Worktree) addOrUpdateFileToIndex(filename string, h plumbing.Hash) error {
+	idx, err := w.r.Storer.Index()
+	if err != nil {
+		return err
+	}
+
+	_, err = idx.Entry(filename)
+	if err == index.ErrEntryNotFound {
+		err = w.doAddFileToIndex(idx, filename)
+	}
+
+	if err != nil {
+		return err
+	}
+
+	err = w.doUpdateFileToIndex(idx, filename, h)
+	if err != nil {
+		return err
+	}
+
+	return w.r.Storer.SetIndex(idx)
+}
+
+func (w *Worktree) doAddFileToIndex(idx *index.Index, filename string) error {
+	idx.Entries = append(idx.Entries, index.Entry{
+		Name: filename,
+	})
+
+	return nil
+}
+
+func (w *Worktree) doUpdateFileToIndex(idx *index.Index, filename string, h plumbing.Hash) error {
```

Comment:
that is performing the action with a function is a name close to this one without the do